### DISCUSSION TITLE
Best practices for development of scrips in Tasks

### DISCUSSION THREAD

**Discussion by afrittoli** [tektoncd/catalog#715] [General]
> Follow-up to a [slack discussion](https://tektoncd.slack.com/archives/CK3HBG7CM/p1619508662150000).

Today we often embed scripts in our task definitions, whether it's [python](https://github.com/tektoncd/catalog/blob/85435f9aa4eb10a7ae7d48d96ad57b39e2bcf394/task/github-set-status/0.2/github-set-status.yaml#L93-L127) or [shell](https://github.com/tektoncd/catalog/blob/85435f9aa4eb10a7ae7d48d96ad57b39e2bcf394/task/git-batch-merge/0.2/git-batch-merge.yaml#L85-L143) or else.

This approach has several advantages:
- it keeps the task definition in one place, making the task more readable
- it helps with things like supply chain security - see [Tekton Chains](https://github.com/tektoncd/chains) for instance
- it helps with deployment and lifecycle of tasks

And a few disadvantages too:
- it promotes bad development practices, as it makes it harder to lint and test the code in the task. We test the Task E2E, but we don't really run unit tests on the scripts within the task
- editing scripts with YAML is not ideal, because of the extra indentation. IDE language tools won't normally work as the IDE treats the YAML field as a text blob

How do we solve this? 

Having the catalog gives us an opportunity for maintaining the advantages of the current approach, and solve its disadvantages at the same time, because we have a clearly defined structure in place we can rely on, which means we can develop tools to improve the development experience.

**Comment by afrittoli** [comment]
Possible solution:
- develop scripts in dedicated subfolder: task/<name>/<version>/<script-xyz>/, along with unit tests
- setup CI so that it runs unit tests and linting on all script folders in a PR
- develop a task template as task/<name>/<version>/<name>.template
- provide a tool that can compile the task template + script folders into a finished task
- store the finished task in git as usual
- setup CI so that it checks that the generated task is up to date with the template and scripts

This means that we can now develop and test scripts normally, and at the same time have them embedded in tasks - so that it's still possible to retrieve / install a full task with a single command / from a single file.
The disadvantage I see in this approach is the extra step in the development workflow - the task needs now to be compiled before it can be tested / re-deployed.

**Comment by chmouel** [comment]
Related 👍🏻 
https://blog.chmouel.com/2020/07/28/tekton-yaml-templates-and-script-feature/

**Comment by chmouel** [comment]
Note that some editor are capable of mixing languages, for example with goland/intellij you can define a "fragment" as another programming language which works quite well. 

I think for Emacs there is https://github.com/polymode/polymode but I haven't figured how to set this up..

![image](https://user-images.githubusercontent.com/98980/118358846-ff623680-b580-11eb-95e2-376db6871505.png)
